File: app/src/main/java/com/patientapp/health/data/DailyFormRepository.java
package com.patientapp.health.data;

import androidx.lifecycle.MutableLiveData;

import com.google.firebase.Timestamp;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.ListenerRegistration;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DailyFormRepository {

    private final FirebaseFirestore firestore;

    public DailyFormRepository() {
        this.firestore = FirebaseFirestore.getInstance();
    }

    public void submitForm(String patientId, String doctorId, double temperature,
                           String symptoms, int painLevel, boolean hasOtherSymptoms,
                           String otherSymptomsDescription, boolean tookMedicine,
                           String medicineDescription, Callback<Void> callback) {
        Map<String, Object> data = new HashMap<>();
        data.put("patientId", patientId);
        data.put("doctorId", doctorId);
        data.put("bodyTemperature", temperature);
        data.put("symptomsDescription", symptoms);
        data.put("painLevel", painLevel);
        data.put("hasOtherSymptoms", hasOtherSymptoms);
        data.put("otherSymptomsDescription", otherSymptomsDescription);
        data.put("tookMedicine", tookMedicine);
        data.put("medicineDescription", medicineDescription);
        data.put("submittedAt", Timestamp.now());

        firestore.collection(DailyForm.COLLECTION).add(data)
                .addOnSuccessListener(ref -> callback.onSuccess(null))
                .addOnFailureListener(callback::onError);
    }

    public ListenerRegistration observeFormsByPatient(String patientId,
                                                      MutableLiveData<List<DailyForm>> liveData) {
        return firestore.collection(DailyForm.COLLECTION)
                .whereEqualTo(DailyForm.FIELD_PATIENT_ID, patientId)
                .addSnapshotListener((snapshot, error) -> {
                    if (error != null || snapshot == null) {
                        liveData.setValue(new ArrayList<>());
                        return;
                    }
                    List<DailyForm> forms = parseForms(snapshot.getDocuments());
                    liveData.setValue(forms);
                });
    }

    public ListenerRegistration observeFormsByDoctor(String doctorId,
                                                     MutableLiveData<List<DailyForm>> liveData) {
        return firestore.collection(DailyForm.COLLECTION)
                .whereEqualTo(DailyForm.FIELD_DOCTOR_ID, doctorId)
                .addSnapshotListener((snapshot, error) -> {
                    if (error != null || snapshot == null) {
                        liveData.setValue(new ArrayList<>());
                        return;
                    }
                    List<DailyForm> forms = parseForms(snapshot.getDocuments());
                    liveData.setValue(forms);
                });
    }

    private List<DailyForm> parseForms(List<DocumentSnapshot> documents) {
        List<DailyForm> forms = new ArrayList<>();
        for (DocumentSnapshot doc : documents) {
            DailyForm form = DocumentParser.toDailyForm(doc);
            if (form != null) forms.add(form);
        }
        Collections.sort(forms, (a, b) -> {
            if (a.getSubmittedAt() == null && b.getSubmittedAt() == null) return 0;
            if (a.getSubmittedAt() == null) return 1;
            if (b.getSubmittedAt() == null) return -1;
            return b.getSubmittedAt().compareTo(a.getSubmittedAt());
        });
        return forms;
    }
}

File: app/src/main/java/com/patientapp/health/data/DailyFormRepository.kt
package com.patientapp.health.data

import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

/**
 * Firestore repository for daily health forms.
 * - Submit a form (patientId, doctorId, temperature, symptoms, painLevel).
 * - Real-time flows: forms by patient, forms by doctor.
 */
class DailyFormRepository {

    private val firestore by lazy { FirebaseFirestore.getInstance() }

    suspend fun submitForm(
        patientId: String,
        doctorId: String,
        temperature: Double,
        symptoms: String,
        painLevel: Int,
        hasOtherSymptoms: Boolean,
        otherSymptomsDescription: String,
        tookMedicine: Boolean,
        medicineDescription: String
    ): Result<Unit> {
        return try {
            val data = hashMapOf(
                "patientId" to patientId,
                "doctorId" to doctorId,
                "bodyTemperature" to temperature,
                "symptomsDescription" to symptoms,
                "painLevel" to painLevel,
                "hasOtherSymptoms" to hasOtherSymptoms,
                "otherSymptomsDescription" to otherSymptomsDescription,
                "tookMedicine" to tookMedicine,
                "medicineDescription" to medicineDescription,
                "submittedAt" to Timestamp.now()
            )
            firestore.collection(DailyForm.COLLECTION).add(data).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    fun getFormsByPatient(patientId: String): Flow<List<DailyForm>> = callbackFlow {
        val listener = firestore.collection(DailyForm.COLLECTION)
            .whereEqualTo(DailyForm.FIELD_PATIENT_ID, patientId)
            .orderBy(DailyForm.FIELD_SUBMITTED_AT, com.google.firebase.firestore.Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    trySend(emptyList())
                    return@addSnapshotListener
                }
                val forms = snapshot?.documents?.mapNotNull { it.toDailyForm() } ?: emptyList()
                trySend(forms)
            }
        awaitClose { listener.remove() }
    }

    fun getFormsByDoctor(doctorId: String): Flow<List<DailyForm>> = callbackFlow {
        val listener = firestore.collection(DailyForm.COLLECTION)
            .whereEqualTo(DailyForm.FIELD_DOCTOR_ID, doctorId)
            .orderBy(DailyForm.FIELD_SUBMITTED_AT, com.google.firebase.firestore.Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    trySend(emptyList())
                    return@addSnapshotListener
                }
                val forms = snapshot?.documents?.mapNotNull { it.toDailyForm() } ?: emptyList()
                trySend(forms)
            }
        awaitClose { listener.remove() }
    }
}

private fun com.google.firebase.firestore.DocumentSnapshot.toDailyForm(): DailyForm? {
    return try {
        DailyForm(
            id = id,
            patientId = getString("patientId") ?: "",
            doctorId = getString("doctorId") ?: "",
            bodyTemperature = (get("bodyTemperature") as? Number)?.toDouble() ?: 0.0,
            symptomsDescription = getString("symptomsDescription") ?: "",
            painLevel = (get("painLevel") as? Number)?.toInt() ?: 0,
            hasOtherSymptoms = getBoolean("hasOtherSymptoms") ?: false,
            otherSymptomsDescription = getString("otherSymptomsDescription") ?: "",
            tookMedicine = getBoolean("tookMedicine") ?: false,
            medicineDescription = getString("medicineDescription") ?: "",
            submittedAt = getTimestamp("submittedAt")
        )
    } catch (e: Exception) {
        null
    }
}
